tools/topology/topology2/include/components/host-copier.conf

```diff
@@ -0,0 +1,140 @@
+#
+# Host gateway copier widget
+#
+# A host gateway copier widget. All attributes defined herein are namespaced
+# by alsatplg to "Object.Widget.host-copier.N.attribute_name"
+#
+# Usage: this component can be used by instantiating it in the parent object. i.e.
+#
+# 	Object.Widget.host-copier."N" {
+#		pcm_id	1
+#	}
+#
+# Where N is the unique instance number for the host-copier object within the same alsaconf node.
+
+Class.Widget."host-copier" {
+	#
+	# Pipeline ID for the copier object
+	#
+	DefineAttribute."index" {}
+
+	/* PCM ID that the host-copier is bound with */
+	DefineAttribute."pcm_id" {}
+
+	#
+	# Copier object instance
+	#
+	DefineAttribute."instance" {}
+
+	#include common component definition
+	<include/components/widget-common.conf>
+
+	#
+	# host-copier component UUID
+	#
+	DefineAttribute."uuid" {
+		type "string"
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.uuid"
+	}
+
+	#
+	# Bespoke Attribute Definitions for host copiers
+	#
+
+	DefineAttribute."direction" {
+		type "string"
+		token_ref	"sof_tkn_dai.word"
+		constraints {
+			!valid_values [
+				"playback"
+				"capture"
+			]
+			!tuple_values [
+				0
+				1
+			]
+		}
+	}
+
+	#
+	# cycles per chunk processed
+	#
+	DefineAttribute."cpc" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.word"
+	}
+
+	DefineAttribute."bss_size" {}
+
+	DefineAttribute."is_pages" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.word"
+	}
+
+	DefineAttribute."node_type" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_copier.word"
+		constraints {
+			!valid_values [
+				$HDA_HOST_OUTPUT_CLASS # HD/A host output (-> DSP)
+				$HDA_HOST_INPUT_CLASS # HD/A host input (<- DSP)
+				$HDA_HOST_INOUT_CLASS # HD/A host input/output (rsvd for future use)
+			]
+		}
+	}
+
+	#
+	# Deep buffer size in ms. Applicable only for host copiers
+	#
+	DefineAttribute."deep_buffer_dma_ms" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_copier.word"
+	}
+
+	attributes {
+		#
+		# The host copier widget name would be constructed using the copier pcm_id and
+		# direction. For ex: "host-copier.1.playback".
+		#
+		!constructor [
+			"pcm_id"
+			"direction"
+		]
+
+		#
+		# mandatory attributes that must be provided when the buffer class is instantiated
+		#
+		!mandatory [
+			"num_input_pins"
+			"num_output_pins"
```

Comment:
i've removed the default values and mandatory attributes for both. num_input_pins and num_output_pins come into the picture only where there're more than 1. Otherwise, they're just ignored by the kernel driver and in all our topologies today we dont have more than 1 of these for playback/capture.

Reply:
ANd I was wrong about this. So I've fixed all host-copiers to have the right number of input/output pins based on direction